Add parliamentary term to hansard questions and answers tables

deanmalan opened the pull request:
This will help us identify unique questions and answers when importing them from PMG.

[Pivotal](https://www.pivotaltracker.com/n/projects/2397264/stories/171251394)

deanmalan committed:
Enable the hansard scraper tests

deanmalan committed:
Import attr in hansard tests

deanmalan committed:
Calculate term when importing questions

deanmalan committed:
Fix can't compare date with datetime

deanmalan committed:
Fix get_term_from_date test

deanmalan committed:
Fix local variable 'question' referenced before assignment

deanmalan committed:
Add ParliamentaryTerm model and migration

deanmalan committed:
Create parliamentary terms in migration

deanmalan committed:
Add term column to Question model

deanmalan committed:
Make term date and and str function

deanmalan committed:
Add get_term_from_date method to ParliamentaryTerm

deanmalan committed:
Add migration to create terms for questions

deanmalan committed:
Use get instead of first to get_term

deanmalan committed:
Add 24th parliament

deanmalan committed:
Handle question without term

deanmalan committed:
Make Question term non-nullable

deanmalan committed:
Add term to Answer

deanmalan committed:
Add migration to calculate terms for answers

deanmalan committed:
Make Answer term non-nullable

deanmalan committed:
Include south_africa_parliamentaryterm table with db export

deanmalan committed:
Include term in question and answer unique indices

deanmalan committed:
Get term for question in hansard_q_and_a_scraper

deanmalan committed:
Use term instead of number in test

deanmalan committed:
Fix q and a test

deanmalan commented on pombola/za_hansard/migrations/0007_create_terms_for_questions.py:
```diff
@@ -0,0 +1,37 @@
+
+
+
+def forwards_func(apps, schema_editor):
```
I was thinking we should first delete all of the data from the questions and answers tables because some of the data is incorrect since we've been identifying duplicates incorrectly. So this migration won't be needed since the will be nothing in the table, but I added it just in case.

jbothma replied:
sjoe I'm nervous of deleting all questions in case the code can't import them as effectively as before because of changes over time. Can we rather just delete those with duplicates?

deanmalan replied:
Good idea :+1: I'll write a script for that.

deanmalan commented on pombola/za_hansard/migrations/0008_auto_20200218_1632.py:
```diff
@@ -0,0 +1,19 @@
+        migrations.AlterField(
+            model_name='question',
+            name='term',
+            field=models.ForeignKey(to='south_africa.ParliamentaryTerm'),
```
This migration sets the `term` field to required after running the data migration above to populate the field.

deanmalan committed:
Enable more tests

jbothma commented on pombola/south_africa/migrations/0004_parliamentaryterm.py:
```diff
@@ -0,0 +1,22 @@
+        migrations.CreateModel(
+            name='ParliamentaryTerm',
+            fields=[
+                ('number', models.IntegerField(serialize=False, primary_key=True)),
```
I don't think we should mix the DB notion of primary key with the parliamentary term number - it's generally considered bad practise even though this is a trivial example.
Can we leave django to make an `id` field that's primary key, and create `number` as a separate `unique` field?

jbothma commented on pombola/south_africa/models.py:
```diff
@@ -29,4 +31,20 @@ class AttendanceForOrganisationToggle(Organisation):
+
+
+class ParliamentaryTerm(models.Model):
+    number = models.IntegerField(primary_key=True) # e.g. 25 for the 25th parliament
```
the comment can sommer be `help_text` to the field

jbothma commented on pombola/za_hansard/management/commands/za_hansard_q_and_a_scraper.py:
```diff
@@ -344,8 +345,26 @@ def handle_api_question_and_reply(self, data):
+            question_date = datetime.strptime(data['date'], "%Y-%m-%d").date()
+        except Exception:
+            self.stderr.write(
+                'Could not parse the date for question {}. Skipping.'.\
```
I suspect that these failures won't be seen til we run it manually.
If we re-raise an exception with our better message, will that end up in the source parsing log we recently added?

We won't get mailed about this unless we let it exit with nonzero status - I think that's a bit overkill, but if it's easy to see this in admin, that'd be nice.

deanmalan replied:
You're right, we won't see the errors unless we exit with a non-zero status. It won't end up in the source parsing log either because this scripts fetches questions directly from the PMG questions API instead of parsing source documents. I could add something like a questions parsing log model?

jbothma replied:
yeah let's do that and make it visible in Admin please. If it's easy, also send an error mail at the end if there are new failures logged - not sure if it's easy to identify errors seen before - let's timebox this to half a day? if it's gonna take more time than that I don't think it's worth it right now

jbothma commented on pombola/za_hansard/management/commands/za_hansard_q_and_a_scraper.py:
```diff
@@ -344,8 +345,26 @@ def handle_api_question_and_reply(self, data):
+            term = ParliamentaryTerm.get_term_from_date(question_date)
+        except Exception:
+            self.stderr.write(
+                'Could not determine the parliamentary term for question {}. Skipping.'.\
```
same here

deanmalan committed:
Add ParliamentaryTerm number help_text

jbothma commented on pombola/za_hansard/migrations/0010_create_terms_for_answers.py:
```diff
@@ -0,0 +1,35 @@
+
+def forwards_func(apps, schema_editor):
+    # For every Answer, calculate it's term
+    for answer in Answer.objects.all():
```
why do answers have a term as well? Does an answer's term not depend on the question's term? Can an answer exist without a term?
 I think this duplication of the field can lead to weirdness.

deanmalan replied:
I added it because answers are also checked for existing answers using the number, house and year:https://github.com/OpenUpSA/pombola/blob/master/pombola/za_hansard/management/commands/za_hansard_q_and_a_scraper.py#L429 and because the answers table also has the unique together constraint between the number, house and year: https://github.com/OpenUpSA/pombola/blob/master/pombola/za_hansard/models.py#L289-L294

I agree that it doesn't make much sense to have those columns in the answers column though. I think it might be there because we used to get answers from multiple sources, but now we only get it from PMG and PMG already tells us which question every answer is linked to.

If we don't add the term to the answers table, we'll have to remove the unique together constraint. It might also make sense then to completely remove the number, house and year columns too.

jbothma replied:
ah right. uhm. leave it this way for now then. not sure we need to clean it all up right now, especially after I've delayed this so long already. Let's deal with it when we see problems caused by it. I don't think it's hurting anyone now, is it?

jbothma commented on pombola/za_hansard/models.py:
```diff
@@ -285,20 +286,16 @@ class Answer (models.Model):
+            ('dp_number', 'house', 'year', 'term'),
         )
 
-        # FIXME - When we have Django 1.5 we can have these indices...
```
:)

deanmalan committed:
Add primary key to Parliamentary Term

deanmalan committed:
Merge branch 'master' into delena-hansard-parl-term

deanmalan committed:
Add command to delete questions that were wrongly identified as duplicates

deanmalan committed:
Cater for multiple duplicate questions

Multiple duplicates are impossible with the old db constraint, but this
is just to be future compatible.

deanmalan committed:
Merge remote-tracking branch 'origin/master' into delena-hansard-parl-term

deanmalan committed:
Document the za_hansard_q_and_a_delete_duplicates command

deanmalan committed:
Add date in exception message

deanmalan committed:
Remove duplicate import from test

deanmalan committed:
Add Q&A error messages to list and exit

deanmalan committed:
Remove FIXME comment because we don't really use that model much anymore

deanmalan committed:
Add QuestionParsingError model

deanmalan committed:
Save question parsing errors to model

deanmalan committed:
Add test for question errors

deanmalan committed:
Add question parsing error table to ignore

deanmalan committed:
Fix question parsing error test

deanmalan committed:
Make the error_message field larger

deanmalan committed:
Remove unused question scraping code

deanmalan committed:
Add question parsing error to admin site

deanmalan committed:
Merge branch 'master' into delena-hansard-parl-term

deanmalan committed:
Merge branch 'master' into delena-hansard-parl-term
